typing/typedecl_separability.ml

```diff
@@ -0,0 +1,668 @@
+    let handle_equation param_instance (acc, context) =
+      (* In the theory, GADT equations are of the form
+           ('a = <ty>)
+         for each type parameter 'a of the type constructor. For each
+         such equation, we should "strengthen" the current context in
+         the following way:
+         - if <ty> is another variable 'b,
+           the mode of 'a is set to the mode of 'b,
+           and 'b is set to Ind
+         - if <ty> is a type expression whose variables are all Ind,
+           set 'a to Ind and discard the equation
+         - otherwise (one of the variable of 'b is not Ind),
+           set 'a to Deepsep and set all variables of <ty> to Ind
+
+         In practice, type parameters are determined by their position
+         in a list, they do not necessarily have a corresponding type variable.
+         Instead of "setting 'a" in the context as in the description above,
+         we build a list of modes by repeated consing into
+         an accumulator variable [acc], setting existential variables
+         to Ind as we go. *)
+      let param_instance = Ctype.repr param_instance in
+      let get context var =
+        try TVarMap.find var context with Not_found -> Ind in
+      let set_ind context var =
+        TVarMap.add var Ind context in
+      let is_ind context var = match get context var with
+        | Ind -> true
+        | Sep | Deepsep -> false in
+      match param_instance.desc with
+      | Tvar text ->
+          let var = {text; id = param_instance.Types.id} in
+          (get context var) :: acc, (set_ind context var)
+      | _ ->
+          let instance_exis = free_variables param_instance in
+          if List.for_all (is_ind context) instance_exis then
+            Ind :: acc, context
+          else
+            Deepsep :: acc, List.fold_left set_ind context instance_exis
+    in
+    let mode_signature, context =
+      (* fold_right here is necessary to get the mode
+         consed to the accumulator in the right order *)
+      List.fold_right handle_equation parameters ([], context) in
+    (* After all variable determined by the parameters have been set to Ind
+       by [handle_equation], all variables remaining in the context are
+       purely existential and should not require a stronger mode than Ind. *)
+    let check_existential evar mode =
+      if rank mode > rank Ind then
+        raise (Error (decl_loc, Non_separable_evar evar.text))
+    in
+    TVarMap.iter check_existential context;
+    mode_signature
+
+(** [check_def env def] returns the signature required
+    for the type definition [def] in the typing environment [env].
+
+    The exception [Not_separable] is raised if we discover that
+    no such signature exists -- the definition will always be invalid. This
+    only happens when the definition is marked to be unboxed. *)
+let check_def
+  : Env.t -> type_definition -> Sep.signature
+  = fun env def ->
+  let unboxed = def.type_unboxed.unboxed in
```

Comment:
Since every occurrence of `unboxed` is negated (`not unboxed`) I'd prefer to see this as
```ocaml
let boxed = not def.type_unboxed.unboxed
```
and have assertions without the double negations below:
```ocaml
assert boxed;
```